// @redstone-finance/sdk/dist/src/contracts/ContractParamsProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractParamsProvider = void 0;
const bytes_1 = require("@ethersproject/bytes");
const utf8_1 = require("@ethersproject/strings/lib/utf8");
const utils_1 = require("ethers/lib/utils");
const index_1 = require("../index");
class ContractParamsProvider {
    requestParams;
    constructor(requestParams) {
        this.requestParams = requestParams;
    }
    async getPayloadHex(withPrefix = true) {
        return ((withPrefix ? "0x" : "") + (await this.requestPayload(this.requestParams)));
    }
    async getPayloadData() {
        return Array.from((0, utils_1.arrayify)(await this.getPayloadHex(true)));
    }
    getHexlifiedFeedIds() {
        return this.getDataFeedIds().map((feed) => (0, bytes_1.hexlify)((0, utf8_1.toUtf8Bytes)(feed)));
    }
    getDataFeedIds() {
        return this.requestParams.dataFeeds;
    }
    async requestPayload(requestParams) {
        return await (0, index_1.requestRedstonePayload)(requestParams);
    }
}
exports.ContractParamsProvider = ContractParamsProvider;
